Mfdeakin sandia/separate net cdf c fortran

mfdeakin-sandia opened the pull request:
Adds support for separate NetCDF C and Fortran library installations.

Test suite: scripts_regression_tests
Test baseline: NA
Test namelist changes: NA
Test status: Pass

Fixes #907 

User interface changes?: To use separate NetCDF C and Fortran libraries, specify NETCDF_C_PATH and NETCDF_FORTRAN_PATH in config_compilers.xml before creating the case.

Code review:

jedwards4b commented on cime_config/acme/machines/Makefile:
```diff
@@ -185,14 +185,40 @@ ifndef AR
-  ifndef LIB_NETCDF
-    LIB_NETCDF:=$(NETCDF_PATH)/lib
+# To prevent breaking cleaning and building components which don't require NetCDF directly
+ifeq ($(if $(or $(filter pio, $(subst /,  , $(MAKECMDGOALS))), $(filter exec_se, $(subst /,  , $(MAKECMDGOALS)))), true, false), true)
```
Can you explain what you are doing here - I'm sure we can find a less complicated method.

mfdeakin-sandia replied:
This detects if we're building either a PIO library or the coupled executable. I'd definitely prefer a simpler method; one thought I had was defining another environment variable for those build steps, but that involves modifying the Python.
The substitution is needed for the pio library because the target is actually a path which depends on the case name; so determining if pio is being built requires searching the string for it.
The substitution isn't actually needed for the coupled executable, so I can remove that.

jedwards4b replied:
The variable MODEL in the makefile is what you want.   When pio is built MODEL=PIO1 or PIO2, when the executable is built MODEL = CIME_MODEL (acme or cesm)

jedwards4b commented on cime_config/acme/machines/Makefile:
```diff
@@ -327,8 +352,10 @@ endif
-ifdef INC_NETCDF
+ifeq ($(NETCDF_SEPARATE), false)
   INCLDIR += -I$(INC_NETCDF)
+else ifeq ($(NETCDF_SEPARATE), true)
```
we don't need the comparison here - just else is fine

mfdeakin-sandia replied:
Won't that fail if NETCDF_SEPARATE isn't defined?

jedwards4b replied:
Okay, I see the point.

jedwards4b commented on cime_config/acme/machines/Makefile:
```diff
@@ -450,8 +491,13 @@ endif
+  ifeq ($(NETCDF_SEPARATE), false)
+    SLIBS := -L$(LIB_NETCDF) -lnetcdf
+  else ifeq ($(NETCDF_SEPARATE), true)
+    SLIBS := -L$(LIB_NETCDF_C) -L$(LIB_NETCDF_FORTRAN) -lnetcdf
```
Actually you are only linking the netcdf c library here - the netcdf fortran library is netcdff.
You should be able to modify these SLIBS lines to have both, and I believe that it will then work for both shared and static links
-lnetcdf -lnetcdff

mfdeakin-sandia replied:
I'm linking the same libraries as the old code, so I'm not certain it's necessary?

jedwards4b replied:
No I'm pretty sure this is wrong - it's only working on skybridge because of the change you made in config_compilers.xml

jedwards4b commented on cime_config/acme/machines/Makefile:
```diff
@@ -659,8 +714,13 @@ db_flags:
 	$(LD) -o $@ test_nc.o -L$(LIB_NETCDF) -lnetcdf $(LDFLAGS)
+else ifeq ($(NETCDF_SEPARATE), true)
+test_nc: test_nc.o
+	$(LD) -o $@ test_nc.o -L$(LIB_NETCDF_C) -lnetcdf $(LDFLAGS)
```
this should be -lnetcdf -lnetcdff

mfdeakin-sandia commented:
I've implemented the requested changes. Anything else?

mfdeakin-sandia commented:
Also, @jgfouca, you might want to verify these changes on Skybridge yourself as well; as I'm not using the sems modules and have issues loading them.

jgfouca commented:
@mfdeakin-sandia as long as you're using the environment defined by config_machines.xml, should be fine.

jedwards4b commented on cime_config/acme/machines/Makefile:
```diff
@@ -185,14 +185,41 @@ ifndef AR
-    LIB_NETCDF:=$(NETCDF_PATH)/lib
+# To prevent breaking cleaning and building components which don't require NetCDF directly
+# NetCDF is only required for building the coupled model and the PIO library
+ifeq ($(if $(or $(findstring PIO, $(MODEL)), $(findstring CIME_MODEL, $(MODEL))), true, false), true)
```
CIME_MODEL is not going to be found in $(MODEL) cesm or acme is.  cesm in the cesm Makefile, acme in the acme one.

mfdeakin-sandia commented:
@jgfouca The issue is that I changed the skybridge entry in config_compilers.xml, as it was attempting to use nf-config which doesn't exist with the netcdf-intel module I'm using.
@jedwards4b $(MODEL) is reported as driver when I print it in acme.bldlog, so I've changed it to that instead.
